<a href="https://colab.research.google.com/github/SojeongShin/NLP_2023/blob/Notion/4_1_BOW_%E1%84%80%E1%85%B5%E1%84%87%E1%85%A1%E1%86%AB%E1%84%8B%E1%85%B4_%E1%84%86%E1%85%AE%E1%86%AB%E1%84%89%E1%85%A5%E1%84%87%E1%85%AE%E1%86%AB%E1%84%85%E1%85%B2_(2).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# [자연어처리]
# 4주차(4-1). BOW 기반의 문서 분류(2)


# 202001581 신소정 <- 본인의 학번, 이름으로 변경하여 제출
* **모든 셀 실행** 후 제출하시기 바랍니다.
* **실습 (4-1-1)**이 있습니다.

# 이제 Colab 사용에 모든 수강생분들이 어느정도 익숙해 졌을 거라 생각됩니다!
* (벌써 4주차 입니다. 시간 빠르죠?)
* **키보드**(+마우스)를 조금 더 사용해보도록 하겠습니다.
> ##### 1. **주석을 해제**하거나
> ##### 2. 다시 **따라 쳐보면서** 실습을 진행해 봅시다. (V)
> ##### 중간 중간 **'# ... 이 부분을 완성하시오'** 라는 부분의 코드를 완성해야 합니다.





## 3-2. 수업 내용 리뷰 시작
## 20 뉴스그룹 데이터 준비 및 특성 추출

 http://scikit-learn.org/0.19/datasets/twenty_newsgroups.html

In [6]:
# from sklearn.datasets import fetch_20newsgroups
from sklearn.datasets import fetch_20newsgroups

# 20개의 토픽 중 선택하고자 하는 토픽을 리스트로 생성
# categories = ['alt.atheism', 'talk.religion.misc', 'comp.graphics', 'sci.space']
categories = [ 'alt.atheism', 'talk.religion.misc', 'comp.graphics', 'sci.space' ]

# 학습 데이터셋을 가져옴
# 메일 내용에서 hint가 되는 부분을 삭제 - 순수하게 내용만으로 분류
# newsgroups_train = fetch_20newsgroups(subset=???, # ... 이 부분을 완성하시오
#                                       remove=('headers', 'footers', 'quotes'),
#                                       categories=categories)
newsgroups_train = fetch_20newsgroups(subset='train',
                  remove=('headers', 'footers', 'quotes'),
                  categories = categories)

# 테스트 데이터셋을 가져옴
# newsgroups_test = fetch_20newsgroups(subset=???, # ... 이 부분을 완성하시오
#                                      remove=('headers', 'footers', 'quotes'),
#                                      categories=categories)
newsgroups_test = fetch_20newsgroups(subset='test',
                remove=('headers', 'footers', 'quotes'),
                categories = categories)

print('#Train set size:', len(newsgroups_train.data))
print('#Test set size:', len(newsgroups_test.data))
print('#Selected categories:', newsgroups_train.target_names)
print('#Train labels:', set(newsgroups_train.target))

#Train set size: 2034
#Test set size: 1353
#Selected categories: ['alt.atheism', 'comp.graphics', 'sci.space', 'talk.religion.misc']
#Train labels: {0, 1, 2, 3}


In [7]:
X_train = newsgroups_train.data   # 학습 데이터셋 문서
y_train = newsgroups_train.target # 학습 데이터셋 레이블

X_test = newsgroups_test.data     # 테스트 데이터셋 문서
y_test = newsgroups_test.target   # 테스트 데이터셋 레이블

In [8]:
from sklearn.feature_extraction.text import TfidfVectorizer

# TF-IDF 방식 / feature의 개수는 2000개를 사용
# tfidf = ???(max_features=???, min_df=5, max_df=0.5) # ... 이 부분을 완성하시오
tfidf = TfidfVectorizer(max_features=2000, min_df = 5, max_df=0.5)
X_train_tfidf = tfidf.fit_transform(X_train) # DTM 2043*2000
X_test_tfidf = tfidf.transform(X_test) #DTM 1353*2000
# train set을 변환
# X_train_tfidf = ???  # ... 이 부분을 완성하시오
# test set을 변환
# X_test_tfidf = ???  # ... 이 부분을 완성하시오

In [10]:
X_train_tfidf.shape

(2034, 2000)

In [13]:
# sklearn이 제공하는 MultinomialNB 를 사용
from sklearn.naive_bayes import MultinomialNB

NB_clf = MultinomialNB()
NB_clf.fit(X_train_tfidf, y_train)

# 분류기 선언
# NB_clf = ???  # ... 이 부분을 완성하시오
# tfidf train set을 이용하여 분류기(classifier)를 새로 학습
# NB_clf.fit(???, ???)  # ... 이 부분을 완성하시오

print('Train set score: {:.3f}'.format(NB_clf.score(X_train_tfidf, y_train))) # train set에 대한 예측정확도를 확인 .3f -> 소수점4번째 자리에서 반올림
print('Test set score: {:.3f}'.format(NB_clf.score(X_test_tfidf, y_test))) # test set에 대한 예측정확도를 확인

Train set score: 0.862
Test set score: 0.741


# 3-2. 수업 내용 리뷰 끝
---
# 4-1. 수업 내용 시작



## 4. 로지스틱 회귀분석을 이용한 문서 분류

1: http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html

In [14]:
# 예시 사용

In [16]:
# sklearn이 제공하는 LogisticRegression을 사용
# from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LogisticRegression

LR_clf = LogisticRegression()
LR_clf.fit(X_train_tfidf, y_train)

# count vector에 대해 regression을 해서 NB와 비교
# LR_clf = LogisticRegression() # 분류기 선언
# LR_clf.fit(X_train_tfidf, y_train) # train data를 이용하여 분류기를 학습

print('Train set score: {:.3f}'.format(LR_clf.score(X_train_tfidf, y_train))) # train data에 대한 예측정확도
print('Test set score: {:.3f}'.format(LR_clf.score(X_test_tfidf, y_test))) # test data에 대한 예측정확도

Train set score: 0.930
Test set score: 0.734


train의 정확도는 올라갔지만, test의 정확도는 내려감.
과적합 발생.

어떻게 해결?

단어 수를 줄인다. -> 라쏘 회귀


계수 값이 지나치게 커지는 것을 막는다. -> 릿지 회귀

In [19]:
LR_clf.coef_  # 4개 features에 대한 2000개씩 각각의 확률

array([[-0.29006795,  0.17889691, -0.19347233, ..., -0.4098512 ,
        -0.08720668, -0.11829089],
       [ 0.41971947, -0.15813167,  0.52840436, ..., -0.13105739,
        -0.05545029,  0.27214755],
       [ 0.10885394,  0.1077093 , -0.18031841, ..., -0.27200652,
         0.10836122, -0.07062918],
       [-0.23850546, -0.12847455, -0.15461362, ...,  0.81291511,
         0.03429575, -0.08322748]])

In [20]:
len(-LR_clf.coef_[0])

2000

영향력 있는 단어들 인덱스 순으로 정렬

In [21]:
import numpy as np

top10 = np.argsort(-LR_clf.coef_[0])[:10]
top10

array([ 214, 1486,  281,  216,  913,  914,  498,  215, 1153, 1424])

In [24]:
# 4개의 카테고리로 늘린다면?
# ... 이 부분을 완성하시오 (전체 코드 작성)
categories = newsgroups_train.target_names
for i, category in enumerate (categories):
  top10 = np.argsort(-LR_clf.coef_[i])[:10]
  print(category, top10)

alt.atheism [ 214 1486  281  216  913  914  498  215 1153 1424]
comp.graphics [ 776  866  682  402   48  683  823 1033 1345  374]
sci.space [1659 1172 1251  976 1142 1660 1612  570 1046  695]
talk.religion.misc [ 356  358  923  765  670 1215 1256  830  808  278]


In [32]:
# get_feature_names_out()을 이용하여, feature로 사용된 단어를 가져온다면?
# ... 이 부분을 완성하시오 (전체 코드 작성)
feature_names = tfidf.get_feature_names_out()
categories = newsgroups_train.target_names
for i, category in enumerate (categories):
  top10 = np.argsort(-LR_clf.coef_[i])[:10]
  for t10 in top10:
    print(category, feature_names[t10], -LR_clf.coef_[i][t10])

alt.atheism atheism -1.9271572105637025
alt.atheism religion -1.7883083820415806
alt.atheism bobby -1.7125439607353627
alt.atheism atheists -1.607440719425755
alt.atheism islam -1.5431861049957143
alt.atheism islamic -1.3272006887040395
alt.atheism deletion -1.3132464987573003
alt.atheism atheist -1.275956128350012
alt.atheism motto -1.1981210256793349
alt.atheism punishment -1.1750045853450275
comp.graphics graphics -3.394586981478147
comp.graphics image -2.244408426718166
comp.graphics file -2.0723853360233124
comp.graphics computer -1.8851287331049897
comp.graphics 3d -1.856267977210757
comp.graphics files -1.7003826581687105
comp.graphics hi -1.5812268147236146
comp.graphics looking -1.4031456901927999
comp.graphics points -1.3926060828797573
comp.graphics code -1.3526118650802066
sci.space space -4.635176130371893
sci.space nasa -2.072639212962214
sci.space orbit -1.9634727483486216
sci.space launch -1.6515920698428153
sci.space moon -1.5778454630258478
sci.space spacecraft -1.425

In [27]:
def top10_features(classifier, vectorizer, categories):
    feature_names = np.asarray(vectorizer.get_feature_names_out())
    for i, category in enumerate(categories):
        # 역순으로 정렬하기 위해 계수에 음수를 취해서 정렬 후 앞에서부터 10개의 값을 반환
        top10 = np.argsort(-classifier.coef_[i])[:10]
        # 카테고리와 영향이 큰 특성 10개를 출력
        print("%s: %s" % (category, ", ".join(feature_names[top10])))

In [28]:
top10_features(LR_clf, tfidf, categories) # ... 이 부분을 완성하시오

alt.atheism: atheism, religion, bobby, atheists, islam, islamic, deletion, atheist, motto, punishment
comp.graphics: graphics, image, file, computer, 3d, files, hi, looking, points, code
sci.space: space, nasa, orbit, launch, moon, spacecraft, shuttle, earth, lunar, flight
talk.religion.misc: christian, christians, jesus, god, fbi, objective, order, his, he, blood


### 라쏘 회귀분석(lasso regression)을 이용한 특성 선택(feature selection)

1. http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Lasso.html

In [29]:
lasso_clf = LogisticRegression(penalty='l1', solver='liblinear', C=1) # Lasso는 동일한 LogisticRegression을 사용하면서 매개변수로 지정
lasso_clf.fit(X_train_tfidf, y_train) # train data로 학습

print('#Train set score: {:.3f}'.format(lasso_clf.score(X_train_tfidf, y_train)))
print('#Test set score: {:.3f}'.format(lasso_clf.score(X_test_tfidf, y_test)))

#Train set score: 0.819
#Test set score: 0.724


In [30]:
# 계수(coefficient) 중에서 0이 아닌 것들의 개수를 출력
print('#Used features count: {}'.format(np.sum(lasso_clf.coef_ != 0)), 'out of', X_train_tfidf.shape[1])

#Used features count: 437 out of 2000


In [31]:
top10_features(lasso_clf, tfidf, newsgroups_train.target_names)

alt.atheism: bobby, atheism, atheists, islam, religion, islamic, motto, atheist, satan, vice
comp.graphics: graphics, image, 3d, file, computer, hi, video, files, looking, sphere
sci.space: space, orbit, launch, nasa, spacecraft, flight, moon, dc, shuttle, solar
talk.religion.misc: fbi, christian, christians, christ, order, jesus, children, objective, context, blood


### 릿지 회귀(ridge regression)


In [ ]:
# from sklearn.linear_model import RidgeClassifier

# ridge_clf = RidgeClassifier() # 릿지 분류기 선언
# ridge_clf.fit(X_train_tfidf, y_train) # 학습

# print('Train set score: {:.3f}'.format(ridge_clf.score(X_train_tfidf, y_train)))
# print('Test set score: {:.3f}'.format(ridge_clf.score(X_test_tfidf, y_test)))

In [ ]:
# top10_features(ridge_clf, tfidf, newsgroups_train.target_names)

In [ ]:
# from sklearn.model_selection import train_test_split

# 8:2의 비율로 나누기
# X_train_ridge, X_val_ridge, y_train_ridge, y_val_ridge = train_test_split(X_train_tfidf, y_train, test_size=???, random_state=42) # ... 이 부분을 완성하시오

In [ ]:
# alpha 값을 0.1로 할 경우
# ridge_clf = RidgeClassifier(alpha=0.1) # 릿지 분류기 선언
# ridge_clf.fit(X_train_tfidf, y_train) # 학습

# print('Train set score: {:.3f}'.format(ridge_clf.score(X_train_tfidf, y_train)))
# print('Test set score: {:.3f}'.format(ridge_clf.score(X_test_tfidf, y_test)))

In [ ]:
# validation에 대한 score를 잰다면?
# ... 이 부분을 완성하시오 (전체 코드 작성)

In [ ]:
# alpha를 0.1부터 10까지 0.1씩 증가
# ... 이 부분을 완성하시오 (전체 코드 작성)

In [ ]:
# max_score = 0
# max_alpha = 0
# for alpha in np.arange(0.1, 10, 0.1): # alpha를 0.1부터 10까지 0.1씩 증가
#     ridge_clf = RidgeClassifier(alpha=alpha) # 릿지 분류기 선언
#     ridge_clf.fit(X_train_ridge, y_train_ridge) # 학습
#     score = ridge_clf.score(X_val_ridge, y_val_ridge) # 검정 데이터셋에 대해 정확도를 측정
#     if score > max_score: # 정확도가 이전의 정확도 최대값보다 크면 최대값을 변경한다.
#         max_score = score
#         max_alpha = alpha
# print('Max alpha {:.3f} at max validation score {:.3f}'.format(max_alpha, max_score))

In [ ]:
# 릿지 분류기 선언
# ridge_clf = RidgeClassifier(alpha=???) # ... 이 부분을 완성하시오
# ridge_clf.fit(X_train_tfidf, y_train) # 학습

# print('Train set score: {:.3f}'.format(ridge_clf.score(X_train_tfidf, y_train)))
# print('Test set score: {:.3f}'.format(ridge_clf.score(X_test_tfidf, y_test)))

In [ ]:
# top10_features(ridge_clf, tfidf, newsgroups_train.target_names)

## 5. 결정트리 등을 이용한 기타 문서 분류 방법

1. https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html
2. https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html
3. https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.GradientBoostingClassifier.html

In [ ]:
# DecisionTreeClassifier
# from sklearn.tree import DecisionTreeClassifier

# tree = DecisionTreeClassifier(random_state=7)
# tree.fit(X_train_tfidf, y_train)

# print('#Decision Tree train set score: {:.3f}'.format(tree.score(X_train_tfidf, y_train)))
# print('#Decision Tree test set score: {:.3f}'.format(tree.score(X_test_tfidf, y_test)))

In [ ]:
# RandomForestClassifier
# from sklearn.ensemble import RandomForestClassifier

# forest = RandomForestClassifier(random_state=7)
# forest.fit(X_train_tfidf, y_train)
# print('#Random Forest train set score: {:.3f}'.format(forest.score(X_train_tfidf, y_train)))
# print('#Random Forest test set score: {:.3f}'.format(forest.score(X_test_tfidf, y_test)))

In [ ]:
# GradientBoostingClassifier
# from sklearn.ensemble import GradientBoostingClassifier

# gb = GradientBoostingClassifier(random_state=7)
# gb.fit(X_train_tfidf, y_train)
# print('#Gradient Boosting train set score: {:.3f}'.format(gb.score(X_train_tfidf, y_train)))
# print('#Gradient Boosting test set score: {:.3f}'.format(gb.score(X_test_tfidf, y_test)))

In [ ]:
# gb.feature_importances_

In [ ]:
# sorted(gb.feature_importances_, reverse=True)[:40]

In [ ]:
# sorted_feature_importances = sorted(zip(tfidf.get_feature_names_out(), gb.feature_importances_), key=lambda x: x[1], reverse=True)
# for feature, value in sorted_feature_importances[:40]:
    # print('%s: %.3f' % (feature, value), end=', ')

# [[실습 4-1-1]]
## 지난 시간과 이번 시간에 배운 분류기들의 train/test score를 비교해 보시오.
* 코드 실행 결과로 얻은 score 결과를 복사/붙여넣기 혹은 직접 입력
* 비교를 통해 어떤 분류기의 성능이 가장 좋은지 평가해 보시오.


# (해당 셀을 더블 클릭하여 수정)
1. NB 분류기
Train set score: 0.862
Test set score: 0.741
2. LR 분류기
Train set score: 0.930
Test set score: 0.734
3. LASSO 분류기
Train set score: 0.819
Test set score: 0.724

- train score가 가장 높은 분류기: LR 분류기
- test score가 가장 높은 분류기: NB 분류기

LR 분류기는 과적합(Overfitting)이 과하기에 분류기로 적합하지 않다. 확률이 낮은 변수도 분류를 하기에 정규화의 의미가 성립되기 어렵다.

NB 분류기는 LR 분류기보단 과적합이 덜 일어나지만 LASSO 분류기보다는 더 일어난다.

따라서 LASSO 분류기가 문서 분류에 가장 적합하다고 생각한다. 이 때, 람다를 조절하여 overfitting, underfitting을 조절할 수 있다.

# [[실습 4-1-2]]
# coef_의 결과와 feature_importances_ 결과 비교를 통해, 회귀분석 방법과 트리 기반의 분류 방식의 차이점 알아보기
* coef_의 결과(여러 분류기 중 1개 이상)와 feature_importances_ 결과(여러 분류기 중 1개 이상) 복사/붙여넣기
* 비교를 통해 어떤 차이점이 있는 것 같은지 분석해보시오.

# (해당 셀을 더블 클릭하여 수정)
* coef_의 결과
* feature_importances_ 결과

# [파일] -> [다운로드] -> [.ipynb 다운로드]
# 제출 후 **구글 계정 로그아웃** 잘 하시기 바랍니다.
# 4-1. 수업 내용 끝